docs: rework mdbook (#1036)
coreos/clair: Documentation/concepts/indexing.md

# Indexing

The [Indexer](../reference/indexer.md) service is responsble for "indexing a manifest".

> Indexing involves taking a manifest representing a container image and computing its constinuent parts. The indexer is trying to understand what packages exist in the image, what distribution the image is derived from, and what package repositories are used within the image. Once this information is computed it is persisted in an IndexReport.

Reviewer: ```suggestion
Indexing involves taking a manifest representing a container image and computing its constituent parts. The indexer is trying to discover what packages exist in the image, what distribution the image is derived from, and what package repositories are used within the image. Once this information is computed it is persisted in an IndexReport.
```

The IndexReport is an intermediate datastructure describing the contents of a container image. This report can be fed to a [Matcher](../reference/matcher.md) node for vulnerability analysis.

> 

Reviewer: ```suggestion
The IndexReport is an intermediate data structure describing the contents of a container image. This report can be fed to a [Matcher](../reference/matcher.md) node for vulnerability analysis.

```

## Content Addressability

> ClairV4 treats all manifests and layers as [content addressable](https://en.wikipedia.org/wiki/Content-addressable_storage). In the context of ClairV4 this means once we index a specific manifest we will not index it again unless its required. Likewise with individual layers. This allows a large reduction in work. 

Reviewer: ```suggestion
ClairV4 treats all manifests and layers as [content addressable](https://en.wikipedia.org/wiki/Content-addressable_storage). In the context of ClairV4 this means once we index a specific manifest we will not index it again unless it's required, and likewise with individual layers. This allows a large reduction in work. 
```

> For example consider how many images in a registry may use "ubuntu:artful" as a base layer. It could be a large majority of images if the developers prefer basing their images off ubuntu. Treating the layers and manifests as content addressable means we will only fetch and scan the ubuntu base layer once.

Reviewer: ```suggestion
For example, consider how many images in a registry may use "ubuntu:artful" as a base layer. It could be a large majority of images if the developers prefer basing their images off ubuntu. Treating the layers and manifests as content addressable means we will only fetch and scan the base layer once.
```

There are of course conditions where ClairV4 should re-index a manifest. 

> When ClairV4 update an internal component such as a package scanner and is asked to index a layer it will know to perform the scan with the new package scanner. ClairV4 is smart enough to understand that a component has changed and the IndexReport may be different this time around. 

Reviewer: ```suggestion
When an internal component such as a package scanner is updated, Clair will know to perform the scan with the new package scanner. Clair has enough information to determine that a component has changed and the IndexReport may be different this time around. 
```

> A client can track ClairV4's `index_state` endpoint to understand when an internal component has changed, and subsequent issue re-indexes. See our [api](../howto/api.md) guide to learn how to view our api specification.

Reviewer: ```suggestion
A client can track ClairV4's `index_state` endpoint to understand when an internal component has changed and subsequently issue re-indexes. See our [api](../howto/api.md) guide to learn how to view our api specification.
```

> ## Summary

Reviewer: Why is the summary at the end?

Author: Summary == Conclusion, no? I guess you expect a summary at the start of the doc?

> In summary you should understand that Indexing is the process ClairV4 uses to understand the contents of a container.

Reviewer: ```suggestion
In summary, you should understand that Indexing is the process Clair uses to understand the contents of layers.
```
